File: clojure/fibonacci.clj
(ns fibonacci)

(defn fibo-seq
  "Infinite Fibonacci sequence"
  []
  (->> [1N 1N]               ; We start with the [1 1] pair. 
    (iterate (fn [[a b]]     ; -> ([1 1] [1 2] [2 3] [3 5] [5 8] ...)
               [b (+ a b)])) ; 
    (map first)))            ; -> ( 1     1     2     3     5    ...)

(defn fibo
  "Returns the first n Fibonacci numbers"
  [n]
  (take n (fibo-seq)))

(defn fibo-below
  "Returns all Fibonacci numbers below x"
  [x]
  (take-while #(< % x) (fibo-seq)))

(defn procedural-fibo
  "Prints the first n Fibonacci numbers in a procedural style"
  [n]
  (loop [i 0, a 1N, b 1N]
    (when (< i n)
      (println a)
      (recur (inc i) b (+ a b)))))

File: go/fibonacci.go
// You can run this file at https://play.golang.org/p/ScdzgFleBh
package main

import (
	"fmt"
)

// fibo returns a list of the n first integers in the Fibonacci sequence
func fibo(n int) ([]int, error) {
	if n < 0 {
		return nil, fmt.Errorf("n can't be below 0")
	}
	if n == 0 {
		return nil, nil
	}
	if n == 1 {
		return []int{1}, nil
	}
	xs := []int{1, 1}
	for i := 2; i < n; i++ {
		xs = append(xs, xs[i-1]+xs[i-2])
	}
	return xs, nil
}

func main() {
	xs, err := fibo(20)
	if err != nil {
		fmt.Println(err)
		return
	}
	for i, x := range xs {
		fmt.Printf("%02d: %d\n", i+1, x)
	}

}
